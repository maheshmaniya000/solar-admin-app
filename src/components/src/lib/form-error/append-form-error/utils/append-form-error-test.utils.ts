import { ElementRef, Type } from '@angular/core';
import {
  AbstractControl,
  AsyncValidatorFn,
  ValidationErrors
} from '@angular/forms';
import { SpectatorDirective } from '@ngneat/spectator/jest';
import { isEmpty } from 'lodash-es';
import { Observable, of } from 'rxjs';
import { delay } from 'rxjs/operators';

import { AppendFormErrorDirective } from '../append-form-error.directive';

export class AppendFormErrorTestUtils {
  static readonly asyncRequired: (
    delayInMilliseconds: number
  ) => AsyncValidatorFn =
    (delayInMilliseconds: number) =>
    (control: AbstractControl): Observable<ValidationErrors> =>
      isEmpty(control.value)
        ? of({ required: true }).pipe(delay(delayInMilliseconds))
        : of(null);

  static expectDirectiveToBePresentOn<D extends AppendFormErrorDirective>(
    element: Element,
    type: Type<D>,
    spectator: SpectatorDirective<AppendFormErrorDirective>
  ): void {
    const appendFormErrorDirectiveElement: Element = spectator.query(type, {
      read: ElementRef
    }).nativeElement;
    expect(appendFormErrorDirectiveElement).toBe(element);
  }
}
