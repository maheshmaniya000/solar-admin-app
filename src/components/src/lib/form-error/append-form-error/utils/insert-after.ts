export function insertAfter(newChild: Element, refChild: Element): void {
  refChild.parentNode.insertBefore(newChild, refChild.nextSibling);
}
