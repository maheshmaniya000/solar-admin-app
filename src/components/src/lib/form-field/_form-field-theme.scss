@use '~@angular/material' as mat;

@mixin theme($theme-or-color-config) {
  $color-config: mat.get-color-config($theme-or-color-config);
  $primary: map-get($color-config, primary);
  $warn: map-get($color-config, warn);
  $foreground: map-get($color-config, foreground);
  $is-dark-theme: map-get($color-config, is-dark);
  $underline-color-base: mat.get-color-from-palette($foreground, divider, if($is-dark-theme, 1, 0.87));
  $underline-focused-color: mat.get-color-from-palette($primary, text);

  mat-form-field.mat-form-field-appearance-outline {
    .mat-form-field-subscript-wrapper .mat-hint {
      color: mat.get-color-from-palette($foreground, hint-text);
    }

    &,
    &.mat-form-field-disabled {
      .mat-form-field-outline {
        color: mat.get-color-from-palette($foreground, disabled);
      }
    }

    &,
    &.mat-focused {
      .mat-form-field-outline-thick {
        color: mat.get-color-from-palette($foreground, icon);
      }
    }

    &.mat-form-field-invalid {
      &[doNotAppendFormError] {
        .mat-form-field-outline-thick,
        &.mat-focused .mat-form-field-outline-thick {
          color: $underline-focused-color;
        }
      }

      ::placeholder {
        color: mat.get-color-from-palette($warn);
      }
    }
  }
}
