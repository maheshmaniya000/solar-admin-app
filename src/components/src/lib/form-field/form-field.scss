@use '../icon/icon' as icon;

$horizontal-spacing: 8px;

mat-form-field {
  width: 100%;

  &.mat-form-field {
    display: block;
  }

  &.mat-form-field-appearance-outline {
    ::placeholder {
      transition: none;
      font-style: italic;
    }

    .mat-form-field-wrapper {
      margin: 0;
      padding-bottom: 0;
    }

    .mat-form-field-outline {
      top: 0;
    }

    .mat-form-field-outline-start {
      border-radius: 4px 0 0 4px;
    }

    .mat-form-field-outline-end {
      border-radius: 0 4px 4px 0;
    }

    .mat-form-field-outline-thick {
      .mat-form-field-outline-start,
      .mat-form-field-outline-end,
      .mat-form-field-outline-gap {
        border-width: 1px;
      }
    }

    &[doNotAppendFormError].mat-form-field-invalid {
      &:not(.mat-focused) .mat-form-field-outline {
        opacity: 1;

        &.mat-form-field-outline-thick {
          opacity: 0;
        }
      }
    }

    .mat-form-field-flex {
      align-items: center;
      margin-top: 0;
      padding: 0 $horizontal-spacing;
    }

    .mat-form-field-infix {
      padding: 5px 0;
      border-top: none;
    }

    .mat-form-field-subscript-wrapper {
      position: static;
      top: 0;
      margin-top: 4px;
      padding: 0;
      font-size: 100%;
    }

    &.ng-untouched,
    &.ng-valid,
    &.ng-pending {
      &.no-hint .mat-form-field-subscript-wrapper {
        display: none;
      }
    }

    .mat-form-field-prefix,
    .mat-form-field-suffix {
      position: static;

      .mat-icon {
        @include icon.size(default);
      }
    }

    &:not(.mat-form-field-appearance-legacy) {
      .mat-form-field-prefix,
      .mat-form-field-suffix {
        .mat-icon-button {
          display: block;
          width: auto;
          height: auto;
        }
      }
    }

    .mat-form-field-prefix {
      margin-right: $horizontal-spacing;
    }

    .mat-form-field-suffix {
      margin-left: $horizontal-spacing;
    }

    .mat-form-field-label-wrapper {
      display: none;
    }
  }
}
