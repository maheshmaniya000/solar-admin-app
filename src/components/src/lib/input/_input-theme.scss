@use '~@angular/material' as mat;

@mixin theme($theme-or-color-config) {
  $color-config: mat.get-color-config($theme-or-color-config);
  $primary: map-get($color-config, primary);

  .mat-form-field-invalid[doNotAppendFormError] .mat-input-element {
    caret-color: mat.get-color-from-palette($primary, text);
  }
}
