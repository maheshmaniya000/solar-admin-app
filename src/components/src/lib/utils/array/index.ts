export * from './array-min-max';
export * from './ensure-index-in-array-bounds';
