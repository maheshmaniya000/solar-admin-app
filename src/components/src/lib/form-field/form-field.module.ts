import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { MatFormFieldModule } from '@angular/material/form-field';
import { TranslocoModule } from '@ngneat/transloco';

import { NoHintDirective } from './no-hint/no-hint.directive';

@NgModule({
  declarations: [NoHintDirective],
  imports: [CommonModule, MatFormFieldModule, TranslocoModule],
  exports: [MatFormFieldModule, NoHintDirective]
})
export class FormFieldModule {}
