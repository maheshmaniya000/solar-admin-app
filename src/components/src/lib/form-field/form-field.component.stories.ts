import { APP_BASE_HREF, CommonModule } from '@angular/common';
import { HttpClientModule } from '@angular/common/http';
import { FormsModule } from '@angular/forms';
import { MatButtonModule } from '@angular/material/button';
import { MatIconModule } from '@angular/material/icon';
import { MatInputModule } from '@angular/material/input';
import { BrowserAnimationsModule } from '@angular/platform-browser/animations';
import { moduleMetadata, Story } from '@storybook/angular';

import { TranslocoRootModule } from 'ng-shared/core/transloco-root.module';

import { SgDefaultsModule } from '../defaults/defaults.module';
import { FormErrorModule } from '../form-error/form-error.module';
import { FormFieldModule } from './form-field.module';

export default {
  title: 'Solargis/Components/Form Field',
  decorators: [
    moduleMetadata({
      imports: [
        BrowserAnimationsModule,
        MatInputModule,
        SgDefaultsModule,
        FormFieldModule,
        FormsModule,
        CommonModule,
        MatIconModule,
        MatButtonModule,
        HttpClientModule,
        TranslocoRootModule,
        FormErrorModule
      ],
      providers: [{ provide: APP_BASE_HREF, useValue: '.' }]
    })
  ]
};

export const bottomSpacing: Story = props => ({
  props,
  template: `
    <div class="container">
      <label for="input1">Hint</label>
      <div>
        <mat-form-field>
          <input id="input1" matInput placeholder="placeholder">
          <mat-hint *ngIf="hintVisible">Some hint</mat-hint>
        </mat-form-field>
        <div>Some block to demonstrate form field height without hint.</div>
      </div>

      <label for="input2">Error</label>
      <div>
        <mat-form-field>
          <input id="input2" matInput placeholder="placeholder" ngModel required>
        </mat-form-field>
        <div>
          Some block to demonstrate form field height with dynamic error. Since
          this field is required, you can see its error when its value is empty,
          and you interact with the input at least once (click in and out).
        </div>
      </div>

      <label for="input3">Error + hint</label>
      <div>
        <mat-form-field>
          <input id="input3" matInput placeholder="placeholder" ngModel required>
          <mat-hint *ngIf="hintVisible">Some hint</mat-hint>
        </mat-form-field>
        <div>
          Some block to demonstrate form field height with dynamic error & hint. Since
          this field is required, you can see its error when its value is empty,
          and you interact with the input at least once (click in and out).
        </div>
      </div>
    </div>
  `,
  styles: [
    `
    .container {
      display: grid;
      grid-template-columns: repeat(2, 1fr);
      grid-gap: 10px;
      width: 800px;
    }
  `
  ]
});

bottomSpacing.argTypes = {
  hintVisible: {
    name: 'Hint visible',
    control: { type: 'boolean' },
    defaultValue: false
  }
};

export const disabled: Story = props => ({
  props,
  template: `
    <div class="container">
      <p>
        'Required' error message will be rendered only if this input is touched
         and enabled. To demonstrate that disabled inputs do not display error, first
         interact with it (click in and out) and then set the disabled control to
         'True'.
      </p>
      <mat-form-field>
        <input matInput ngModel [disabled]="disabled" required placeholder="placeholder"/>
      </mat-form-field>
    </div>
  `,
  styles: [
    `
    .container {
      width: 500px;
    }
  `
  ]
});

disabled.argTypes = {
  disabled: {
    name: 'Disabled',
    control: { type: 'boolean' },
    defaultValue: false
  }
};

export const prefixAndSuffix: Story = () => ({
  template: `
    <div class="container">
      <label for="input1">Text</label>
      <mat-form-field>
        <input id="input1" matInput placeholder="placeholder"/>
        <span matPrefix>+421</span>
        <span matSuffix>kWh</span>
      </mat-form-field>
      <label for="input2">Icon / Icon button</label>
      <mat-form-field>
        <input id="input2" matInput placeholder="placeholder"/>
        <mat-icon matPrefix>search</mat-icon>
        <button mat-icon-button color="primary" matSuffix>
          <mat-icon>visibility_off</mat-icon>
        </button>
      </mat-form-field>
    </div>
  `,
  styles: [
    `
    .container {
      display: grid;
      grid-template-columns: repeat(2, 1fr);
      grid-gap: 10px;
      width: 500px;
    }
  `
  ]
});
