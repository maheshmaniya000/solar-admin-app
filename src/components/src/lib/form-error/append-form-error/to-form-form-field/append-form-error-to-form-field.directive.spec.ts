import { ElementRef } from '@angular/core';
import {
  FormControl,
  FormGroup,
  FormsModule,
  ReactiveFormsModule,
  Validators
} from '@angular/forms';
import { MatFormField, MatFormFieldModule } from '@angular/material/form-field';
import { MatInputModule } from '@angular/material/input';
import {
  byText,
  createDirectiveFactory,
  SpectatorDirective
} from '@ngneat/spectator/jest';
import { TranslocoTestingModule } from '@ngneat/transloco';

import { AppendFormErrorTestUtils } from '../utils/append-form-error-test.utils';
import { AppendFormErrorToFormFieldDirective } from './append-form-error-to-form-field.directive';

describe('AppendFormErrorToFormFieldDirective', () => {
  describe('component', () => {
    const createDirective = createDirectiveFactory({
      directive: AppendFormErrorToFormFieldDirective,
      imports: [
        ReactiveFormsModule,
        FormsModule,
        MatFormFieldModule,
        MatInputModule,
        TranslocoTestingModule.forRoot({ langs: {} })
      ],
      shallow: true
    });
    let spectator: SpectatorDirective<AppendFormErrorToFormFieldDirective>;

    describe('directive presence', () => {
      it(`should not attach directive to mat-form-field element that has
 noAppendError attribute on itself`, () => {
        expect(() =>
          createDirective(`
            <mat-form-field doNotAppendFormError>
              <input matInput ngModel/>
            </mat-form-field>
          `)
        ).toThrow();
      });

      it(`should attach directive to mat-form-field element`, () => {
        spectator = createDirective(`
          <mat-form-field>
            <input matInput ngModel/>
          </mat-form-field>
        `);
        AppendFormErrorTestUtils.expectDirectiveToBePresentOn(
          spectator.query(MatFormField, { read: ElementRef }).nativeElement,
          AppendFormErrorToFormFieldDirective,
          spectator
        );
      });
    });

    describe('error message rendering', () => {
      describe('form control', () => {
        it(`should render error message inside mat-form-field
 that contains invalid and touched form control`, () => {
          const formControl = new FormControl(undefined, {
            validators: Validators.required
          });
          spectator = createDirective(
            `
              <mat-form-field>
                <input matInput [formControl]="formControl"/>
              </mat-form-field>
            `,
            {
              hostProps: { formControl }
            }
          );
          spectator.focus('input');
          spectator.blur('input');

          expect(spectator.fixture).toMatchSnapshot();
        });
      });

      describe('form group', () => {
        let secondFormControl: FormControl;
        let thirdFormControl: FormControl;
        let formGroup: FormGroup;
        const minErrorMessage = 'en.common.validation.min';

        beforeEach(() => {
          secondFormControl = new FormControl(4, Validators.min(5));
          thirdFormControl = new FormControl(11, Validators.max(10));
          formGroup = new FormGroup({
            first: new FormControl('some value', Validators.required),
            second: secondFormControl,
            third: thirdFormControl
          });

          spectator = createDirective(
            `
            <mat-form-field>
              <ng-container [formGroup]="formGroup">
                <input matInput formControlName="first"/>
                <input matInput formControlName="second"/>
                <input matInput formControlName="third"/>
              </ng-container>
            </mat-form-field>
          `,
            {
              hostProps: { formGroup }
            }
          );
        });

        it(`should render the highest priority error message of all touched
 invalid form controls in the form group`, () => {
          secondFormControl.markAsTouched();
          spectator.detectChanges();
          expect(spectator.query(byText(minErrorMessage))).toExist();
        });

        it(`should replace error message of the touched invalid form control
 in the form group with the highest priority error message of all touched invalid
 form controls in the form group when the first one becomes valid`, () => {
          secondFormControl.markAsTouched();
          spectator.detectChanges();
          expect(spectator.query(byText(minErrorMessage))).toExist();
          secondFormControl.setValue(5);
          thirdFormControl.markAsTouched();
          spectator.detectChanges();
          expect(spectator.query(byText('en.common.validation.max'))).toExist();
        });

        it(`should render error message of the form group itself when it's
 invalid and touched and contains the highest priority error of all touched invalid
 form controls`, () => {
          formGroup.setValidators(() => ({ integer: true }));
          secondFormControl.setValue(5);
          thirdFormControl.setValue(10);
          secondFormControl.markAsTouched();
          thirdFormControl.markAsTouched();
          spectator.detectChanges();
          expect(
            spectator.query(byText('en.common.validation.integer'))
          ).toExist();
        });
      });
    });
  });
});
