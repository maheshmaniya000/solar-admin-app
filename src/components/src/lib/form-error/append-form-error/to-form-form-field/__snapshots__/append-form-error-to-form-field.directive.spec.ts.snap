// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppendFormErrorToFormFieldDirective component error message rendering form control should render error message inside mat-form-field
 that contains invalid and touched form control 1`] = `
<ng-component
  __ngContext__={[Function LRootView]}
  formControl={[Function FormControl]}
>
  <mat-form-field
    class="mat-form-field mat-primary mat-form-field-type-mat-input mat-form-field-appearance-legacy mat-form-field-can-float mat-form-field-hide-placeholder ng-pristine ng-invalid _mat-animation-noopable mat-form-field-invalid ng-touched"
  >
    <div
      class="mat-form-field-wrapper"
    >
      <div
        class="mat-form-field-flex"
      >
        <div
          class="mat-form-field-infix"
        >
          <input
            aria-invalid="true"
            aria-required="false"
            class="mat-input-element mat-form-field-autofill-control ng-pristine ng-invalid cdk-text-field-autofill-monitored ng-touched"
            matinput=""
          />
          <span
            class="mat-form-field-label-wrapper"
          >
          </span>
        </div>
      </div>
      <div
        class="mat-form-field-underline"
      >
        <span
          class="mat-form-field-ripple"
        />
      </div>
      <div
        class="mat-form-field-subscript-wrapper"
      >
        <div
          class="mat-form-field-hint-wrapper ng-trigger ng-trigger-transitionMessages"
          style="opacity:1;transform:translateY(0%);0:opacity;1:transform;"
        >
          <div
            class="mat-form-field-hint-spacer"
          />
        </div>
        <sg-form-error>
          <mat-error>
             en.common.validation.required 
          </mat-error>
        </sg-form-error>
      </div>
    </div>
  </mat-form-field>
</ng-component>
`;
